### PR TITLE
mpi_t: MPI T specific error handling.

### PR DESCRIPTION
**Pull request by hzhou**
This is related to GitHub issue: https://github.com/pmodels/mpich/issues/2051.

MPI tool interface maybe used before MPI_Init or after MPI_Finalize and they should not abort on errors. This PR ported the current MPIR_ error routines into a set of much simpler MPIT_ routines that do not abort and only returns MPI_T_ERR_ codes.

### CODE REVIEW

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
Test results look good. The freebsd failures is a false positive due to SLURM communication error. We can proceed with review.

**Comment by raffenet** [comment]
Assigned @yfguo for review. This looked OK to me at a glance. The duplication of the memory checking macros is unfortunate, but I think the reason for doing so is necessary, so I'm OK with it.

File: src/include/mpit_err.h [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    if ((index_) < 0 || ((unsigned) index_) >= utarray_len(cvar_table)) { \
+        mpi_errno = MPI_T_ERR_INVALID_INDEX; \
+        goto fn_fail; \
+    }
+
+#define MPIT_ERRTEST_CVAR_HANDLE(handle_) \
+    if ((handle_) == MPI_T_CVAR_HANDLE_NULL) { \
+        mpi_errno = MPI_T_ERR_INVALID_HANDLE; \
+        goto fn_fail; \
+    } else if ((handle_)->kind != MPIR_T_CVAR_HANDLE) { \
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/include/mpit_err.h]: The macro does not require a different style. But I think the new line `else` is the preferred style for the project. @raffenet

File: src/include/mpit_err.h [hunk 2]
```diff
@@ -0,0 +1,106 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2001 by Argonne National Laboratory.
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpit_err.h]: Sure. Fault of copy-pasting. :)

File: src/include/mpit_err.h [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+ *   to check return values and handle errors according to their application logic.
+ *
+ *   To maintain a parallel interface with rest of the MPICH error handling calls,
+ *   corresponding macros in mpir_err.h are ported here with MPIT_ prefix.
+ */
+
+#include <assert.h>
+#define MPIT_Assert(cond_) assert(cond_);
+
+#define MPIT_ERR_POP() \
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpit_err.h]: Yeah I am really having trouble with this one. I routinely split my coding windows into multiple narrow pieces and don't want to set my font too small either. As a result, these long-line padding severely mess up the view to the point of unworkable (for me).

File: src/include/mpit_mem.h [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+#define MPIT_CHKPMEM_DECL(n_) \
+    void *(mpiu_chkpmem_stk_[n_]) = { NULL }; \
+    int mpiu_chkpmem_stk_sp_=0; \
+    const int mpiu_chkpmem_stk_sz_ = n_
+
+#define MPIT_CHKPMEM_COMMIT() \
+    mpiu_chkpmem_stk_sp_ = 0
+
+#define MPIT_CHKPMEM_REAP() \
+    { \
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpit_mem.h]: I have no issue with this.

File: src/include/mpit_err.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpit_err.h]: In maint/code-cleanup.sh line24: `--cuddle-else` :)

[the remark this answers is not in the document]
Comment (yfguo) [reply src/include/mpit_err.h]: OK. Cuddle else then. 😸

File: src/include/mpit_err.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (hzhou) [reply src/include/mpit_err.h]: If you don't mind, I would appreciate if you clean it up for me this time. (Eventually I'll need work out a filter that dynamically reformat this style both way for me. )

**Comment by raffenet** [comment]
@yfguo have the review comments been addressed?

**Comment by raffenet** [comment]
@yfguo pinging again to check the status of the requested changes.

**Comment by pavanbalaji** [comment]
Stale PR from 2018.  Closing.

**Pull request closed without merging**